### DISCUSSION TITLE
USB OTG access for Sync Files

### DISCUSSION THREAD

**Discussion by gmaclennan** [digidem/mapeo-mobile#761] [Ideas]
I am disappointed to report that USB OTG access is not at all easy for syncing to a USB.

**TL;DR** we cannot currently write syncfiles to USB-OTG. The best we can do is write to internal storage and ask the user to move the file to USB, and that requires free space on the device equal to the size of the database.

"OTG" stands for "On The Go" and refers to a USB storage device connected to the Android phone's USB port using a micro-USB to female USB-A adapter or similar.

Android does not mount a USB-OTG drive when you connect it. On my Moto X I can open the system "Files" app and it will mount the drive (I can see it from `adb shell mount`) but I have no read permission to the drive. It does not seem like other apps can mount a USB OTG drive, which makes it impossible to access from the filesystem.

It seems like in a previous version of Android USB OTG was mounted, but at a [different location on every device](https://stackoverflow.com/questions/36086201/scan-otg-mount-point-in-android). According to [this post](https://stackoverflow.com/questions/36862675/android-sd-card-write-permission-using-saf-storage-access-framework#36865620) "In recent version of Android, accessing data outside your application using java.io.File is almost completely deprecated." The only way to access data on the USB is via the [Storage Access Framework (SAF)](https://developer.android.com/guide/topics/providers/document-provider).

There is a [Java module](https://github.com/magnusja/libaums) that facilitates access to USB OTG, and there is a [React-Native module](https://github.com/scizers/react-native-otg-storage) that provides JavaScript bindings to the Native API that enables you to read and write files on the USB. It only seems to support writing of entire files, but the native API in libaums seems to have a streaming API and also an (undocumented) API for random access writes to a file.

I was able to create a [very basic React-Native App](https://github.com/gmaclennan/rnusbotg) that uses `react-native-otg-storage` to list files on the USB device and create a new file.

In conclusion:

- We can read and create files on a USB-OTG from react-native
- USB-OTG devices are never mounted, so the nodejs process cannot read/write files
- It seems like `react-native-otg-storage` would need the entire file to be in memory in the react-native process to write.
- The underlying library [libaums](https://github.com/magnusja/libaums) supports low-level access to files, and should support updating files in-place.

I don't think we can move forwards with this without writing a native (Java) module, building on libaums and react-native-otg-storage. I see two strategies, the first should be relatively easy, the second much harder:

1. We could write a native module with react-native integration that can detect a USB-OTG being connected, detect an existing sync file, and copy it to internal storage, where the nodejs process can sync to it, then back in the react-native process move the sync file back onto the USB. We would need to use the API from libaums to do a direct copy from internal storage to USB without the file moving into the react-native process. The big limit of this is memory: the entire syncfile would need to fit on the phone internal memory, which means that users would only be able to use half of their available space for mapeo-mobile.
2. We could write an implementation of syncfile in Java, excluding the osm-p2p-db bits, that could read and write abstract-storage and provide an HTTP API for reading and writing media in the sync-file in place. The database we could move into internal storage, but this way we could keep the media on the USB drive and save memory. This would need a Java developer who understood the node abstract-storage modules with knowledge of low-level filesystem work (writing tar files at byte offsets etc.)

As far as I can make out, the same restrictions of using the Storage Access Framework also apply to any additional SD card that is used on the phone. This would limit the ability of the nodejs process to read/write files on the external SD card. We would need to do something similar to the above in order to use this storage, or depend on users having Android 6.0+ devices and [adoptable storage](https://source.android.com/devices/storage/adoptable) which makes an external SD card appear as part of internal storage.

**Comment by okdistribute** [comment]
Thanks for doing all of this work gregor, what a rabbit hole. 

What if we extended the current `react-native-otg-storage` to have stream access, so it can write the syncfile in chunks via the `libaums` stream api? The nodejs process can create the syncfile, and then the react-native HTTP client can request chunks and write those chunks to the stream (so only each chunk needs to be in memory, not the whole file...

**Comment by gmaclennan** [comment]
This is a good idea @karissa. I don't know Java but reading through the code it doesn't seem that difficult to do something like this. libaums has an optional http server too, which I haven't investigated. I think the challenge is the two-directional nature of the sync, maybe read the stream writing new files to the db, then create a writestream writing to a new syncfile, finally removing the old one. libaums has code for moving/renaming on device too.

**Comment by ghost** [comment]
Would we still be blocked by

> It does not seem like other apps can mount a USB OTG drive, which makes it impossible to access from the filesystem.

By this, do you mean that it's impossible to mount USB OTG drives by any means via an app? Did I misparse?

**Comment by okdistribute** [comment]
A first pass quick implementation could focus on getting data off of the phone in the case that a WiFi connection is unavailable, into a syncfile by writing it in the user data folder for Mapeo Mobile, from nodejs. 

Can someone who is trained be able to get access to this user data folder to read the contents/copy it to the desktop?

Then we have extra chunks of work beyond this:
- [ ] writing to a USB (exporting data to usb)
- [ ] writing and reading from a USB simultaneously (two-way sync)
- [ ] UX/UI: dataset detection and selection
